### PR TITLE
backend: Compute Test Median

### PR DESCRIPTION
**Pull request by wais-hundekar**
## Notion ticket link

[Compute Test Median](https://www.notion.so/uwblueprintexecs/47bf4727a2444bccad4a91952898b8f3?v=9658436bf9be409e90581631a7155f16&p=e23f791a54b1480a9991abea5f33fcda&pm=s)

## Implementation description
* Adding function to compute median for tests of a given test id

## Steps to test
1. Ensure exists tests are thorough
2. Run existing tests

## What should reviewers focus on?
* Are existing tests thorough enough?
* Should more parts of the pipeline be split into helper functions?

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/services/implementations/__tests__/statisticService.test.ts [hunk 1]
```diff
@@ -103,6 +103,22 @@ describe("mongo statisticService", (): void => {
     expect(actualResult).toEqual(0);
   });
 
+  it("getMedianScoreByTest with multiple submissions", async () => {
```
Comment (carissa-tang) [inline backend/services/implementations/__tests__/statisticService.test.ts]: can you split this into two tests - one with even submissions and with odd submissions?

Comment (wais-hundekar) [reply backend/services/implementations/__tests__/statisticService.test.ts]: Done!

File: backend/testUtils/testSession.ts [hunk 2]
```diff
@@ -185,6 +185,47 @@ export const mockTestSessions: TestSessionRequestDTO[] = [
...
+    results: [
+      mockGradedTestResult,
+      mockGradedTestResult2,
+      mockGradedTestResult3,
+      mockUngradedTestResult,
+      mockGradedTestResult,
+      mockGradedTestResult4,
+    ],
+    accessCode: "1234",
+    startTime: new Date("2021-09-01T09:00:00.000Z"),
```
Comment (carissa-tang) [inline backend/testUtils/testSession.ts]:
```suggestion
    ... mockTestSessions[0],
    results: [
      mockGradedTestResult,
      mockGradedTestResult2,
      mockGradedTestResult3,
      mockUngradedTestResult,
      mockGradedTestResult,
      mockGradedTestResult4,
    ],
```
can we refactor these

Comment (wais-hundekar) [reply backend/testUtils/testSession.ts]: Done!

File: backend/services/implementations/statisticService.ts [hunk 3]
```diff
@@ -97,6 +100,44 @@ class StatisticService implements IStatisticService {
...
+          _id: null,
+          scores: { $push: "$score" },
+        },
+      },
+      {
+        $project: {
+          _id: 0,
+          scores: 1,
+        },
+      },
```
Comment (carissa-tang) [inline backend/services/implementations/statisticService.ts]: don't think this is needed
```suggestion
```

File: backend/testUtils/testSession.ts [hunk 4]
```diff
@@ -185,6 +185,37 @@ export const mockTestSessions: TestSessionRequestDTO[] = [
...
+  },
+  {
+    test: "62c248c0f79d6c3c9ebbea94", // invalid test (will not be created)
+    teacher: mockTeacher.id,
+    school: mockSchoolWithId.id,
+    gradeLevel: 4,
+    results: [mockGradedTestResult],
+    accessCode: "1234",
+    startTime: new Date("2021-09-01T09:00:00.000Z"),
+  },
```
Comment (carissa-tang) [inline backend/testUtils/testSession.ts]: this is same as mockTestSessions[4] right?

**Review by carissa-tang — APPROVED** [review]

**Pull request merged**